File: sdks/java/io/azure-cosmosdb/src/test/scala/org/apache/beam/sdk/io/azure/cosmos/CosmosIOTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.sdk.io.azure.cosmos

import com.azure.cosmos.CosmosClientBuilder
import org.apache.beam.sdk.Pipeline
import org.apache.beam.sdk.io.azure.cosmos.CosmosIOTest.{ CONTAINER, DATABASE, cosmosDBEmulatorContainer }
import org.apache.beam.sdk.testing.PAssert
import org.apache.beam.sdk.transforms.Count
import org.apache.beam.sdk.values.PCollection
import org.bson.Document
import org.junit._
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.testcontainers.containers.CosmosDBEmulatorContainer
import org.testcontainers.utility.DockerImageName
import scribe._

import java.nio.file.Files
import scala.util.Using

@RunWith(classOf[JUnit4])
class CosmosIOTest {
  Logger.root
    .clearHandlers()
    .clearModifiers()
    .withHandler(
      minimumLevel = Some(Level.Debug),
    )
    .replace()

  @Test
  def readFromCosmosCoreSqlApi(): Unit = {
    val read = CosmosIO
      .read()
      .withCosmosEndpoint(cosmosDBEmulatorContainer.getEmulatorEndpoint)
      .withCosmosKey(cosmosDBEmulatorContainer.getEmulatorKey)
      .withQuery(s"SELECT * FROM c")
      .withContainer(CONTAINER)
      .withDatabase(DATABASE)

    val pipeline = Pipeline.create()
    val count: PCollection[java.lang.Long] = pipeline
      .apply(read)
      .apply(Count.globally())

    PAssert.thatSingleton(count).isEqualTo(10)

    pipeline.run().waitUntilFinish()
  }
}

/** Initialization of static fields and methods */
@RunWith(classOf[JUnit4])
object CosmosIOTest {
  Logger.root
    .clearHandlers()
    .clearModifiers()
    .withHandler(
      minimumLevel = Some(Level.Debug),
    )
    .replace()
  private val DOCKER_NAME = "mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:latest"
  private val cosmosDBEmulatorContainer = new CosmosDBEmulatorContainer(DockerImageName.parse(DOCKER_NAME))
  private val DATABASE = "test"
  private val CONTAINER = "test"

  @BeforeClass
  def setup(): Unit = {
    info("Starting CosmosDB emulator")
    cosmosDBEmulatorContainer.start()

    val tempFolder = new TemporaryFolder
    tempFolder.create()
    val keyStoreFile = tempFolder.newFile("azure-cosmos-emulator.keystore").toPath
    val keyStore = cosmosDBEmulatorContainer.buildNewKeyStore
    keyStore.store(Files.newOutputStream(keyStoreFile.toFile.toPath), cosmosDBEmulatorContainer.getEmulatorKey.toCharArray)
    System.setProperty("javax.net.ssl.trustStore", keyStoreFile.toString)
    System.setProperty("javax.net.ssl.trustStorePassword", cosmosDBEmulatorContainer.getEmulatorKey)
    System.setProperty("javax.net.ssl.trustStoreType", "PKCS12")


    info("Creando la data -------------------------------------------------------->")
    val triedCreateData = Using(new CosmosClientBuilder()
      .gatewayMode
      .endpointDiscoveryEnabled(false)
      .endpoint(cosmosDBEmulatorContainer.getEmulatorEndpoint)
      .key(cosmosDBEmulatorContainer.getEmulatorKey)
      .buildClient) { client =>

      client.createDatabase(DATABASE)
      val db = client.getDatabase(DATABASE)
      db.createContainer(CONTAINER, "/id")
      val container = db.getContainer(CONTAINER)
      for (i <- 1 to 10) {
        container.createItem(new Document("id", i.toString))
      }
    }
    if (triedCreateData.isFailure) {
      val throwable = triedCreateData.failed.get
      error("Error creando la data", throwable)
      throw throwable
    }
    info("Data creada ------------------------------------------------------------<")
  }

  @AfterClass
  def close(): Unit = {
    info("Stop CosmosDB emulator")
    cosmosDBEmulatorContainer.stop()
  }
}

File: sdks/java/io/azure-cosmos/src/test/java/org/apache/beam/sdk/io/azure/cosmos/CosmosIOTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.sdk.io.azure.cosmos;

import static org.junit.Assert.assertEquals;

import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.CosmosContainer;
import com.azure.cosmos.CosmosDatabase;
import com.azure.cosmos.examples.common.Families;
import com.azure.cosmos.examples.common.Family;
import com.azure.cosmos.models.CosmosItemRequestOptions;
import com.azure.cosmos.models.PartitionKey;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyStore;
import java.util.ArrayList;
import java.util.List;
import org.apache.beam.sdk.coders.SerializableCoder;
import org.apache.beam.sdk.io.BoundedSource;
import org.apache.beam.sdk.io.azure.cosmos.CosmosIO.BoundedCosmosBDSource;
import org.apache.beam.sdk.io.azure.cosmos.CosmosIO.Read;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.testing.PAssert;
import org.apache.beam.sdk.testing.TestPipeline;
import org.apache.beam.sdk.transforms.Count;
import org.apache.beam.sdk.values.PCollection;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.testcontainers.containers.CosmosDBEmulatorContainer;
import org.testcontainers.utility.DockerImageName;

// See https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/quickstart-java
public class CosmosIOTest {

  private static final String DOCKER_IMAGE_NAME =
      "mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:latest";
  private static final String DATABASE = "AzureSampleFamilyDB";
  private static final String CONTAINER = "FamilyContainer";
  private static final String PARTITION_KEY_PATH = "/lastName";

  private static PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
  private static CosmosDBEmulatorContainer container;
  private static CosmosClient client;

  @BeforeClass
  public static void beforeClass() throws Exception {
    // Create the azure cosmos container and pull image as we use latest
    container = new CosmosDBEmulatorContainer(DockerImageName.parse(DOCKER_IMAGE_NAME));
    container.withImagePullPolicy(imageName -> true);

    // Start the container. This step might take some time...
    container.start();
    TemporaryFolder tempFolder = new TemporaryFolder();
    tempFolder.create();
    Path keyStoreFile = tempFolder.newFile("azure-cosmos-emulator.keystore").toPath();
    KeyStore keyStore = container.buildNewKeyStore();
    keyStore.store(Files.newOutputStream(keyStoreFile), container.getEmulatorKey().toCharArray());
    System.setProperty("javax.net.ssl.trustStore", keyStoreFile.toString());
    System.setProperty("javax.net.ssl.trustStorePassword", container.getEmulatorKey());
    System.setProperty("javax.net.ssl.trustStoreType", "PKCS12");

    CosmosOptions cosmosOptions = pipelineOptions.as(CosmosOptions.class);
    cosmosOptions.setCosmosServiceEndpoint(container.getEmulatorEndpoint());
    cosmosOptions.setCosmosKey(container.getEmulatorKey());

    client =
        new CosmosClientBuilder()
            .gatewayMode()
            .endpointDiscoveryEnabled(false)
            .endpoint(container.getEmulatorEndpoint())
            .key(container.getEmulatorKey())
            .buildClient();

    client.createDatabase(DATABASE);
    CosmosDatabase db = client.getDatabase(DATABASE);
    db.createContainer(CONTAINER, PARTITION_KEY_PATH);
    CosmosContainer container = db.getContainer(CONTAINER);
    List<Family> families = new ArrayList<>();
    families.add(Families.getAndersenFamilyItem());
    families.add(Families.getJohnsonFamilyItem());
    families.add(Families.getSmithFamilyItem());
    families.add(Families.getWakefieldFamilyItem());

    CosmosItemRequestOptions cosmosItemRequestOptions = new CosmosItemRequestOptions();
    for (Family f : families) {
      container.createItem(f, new PartitionKey(f.getLastName()), cosmosItemRequestOptions);
    }
  }

  @AfterClass
  public static void afterClass() {
    container.stop();
    if (client != null) {
      client.close();
    }
  }

  @Rule public TestPipeline pipeline = TestPipeline.fromOptions(pipelineOptions);

  @Test
  public void testEstimatedSizeBytes() throws Exception {
    Read<Family> read =
        CosmosIO.read(Family.class)
            .withContainer(CONTAINER)
            .withDatabase(DATABASE)
            .withCoder(SerializableCoder.of(Family.class));

    BoundedCosmosBDSource<Family> initialSource = new BoundedCosmosBDSource<>(read);
    // Cosmos DB precision is in KB. Inserted test data is ~3KB
    long estimatedSize = initialSource.getEstimatedSizeBytes(pipelineOptions);
    assertEquals("Wrong estimated size", 3072, estimatedSize);
  }

  @Test
  public void testSplit() throws Exception {
    Read<Family> read =
        CosmosIO.read(Family.class)
            .withContainer(CONTAINER)
            .withDatabase(DATABASE)
            .withCoder(SerializableCoder.of(Family.class));

    BoundedCosmosBDSource<Family> initialSource = new BoundedCosmosBDSource<>(read);
    // Cosmos DB precision is in KB. Inserted test data is ~3KB
    List<? extends BoundedSource<Family>> splits = initialSource.split(1024, pipelineOptions);
    assertEquals("Wrong split", 3, splits.size());
  }

  @Test
  public void testRead() {
    PCollection<Family> output =
        pipeline.apply(
            CosmosIO.read(Family.class)
                .withContainer(CONTAINER)
                .withDatabase(DATABASE)
                .withCoder(SerializableCoder.of(Family.class)));

    PAssert.thatSingleton(output.apply("Count", Count.globally())).isEqualTo(4L);
    pipeline.run();
  }

  @Test
  public void testReadWithQuery() {
    PCollection<Family> output =
        pipeline.apply(
            CosmosIO.read(Family.class)
                .withContainer(CONTAINER)
                .withDatabase(DATABASE)
                .withCoder(SerializableCoder.of(Family.class))
                .withQuery(
                    "SELECT * FROM Family WHERE Family.lastName IN ('Andersen', 'Wakefield', 'Johnson')"));

    PAssert.thatSingleton(output.apply("Count", Count.globally())).isEqualTo(3L);
    pipeline.run();
  }
}
